cpp/src/arrow/type_test.cc

```diff
@@ -983,6 +1091,242 @@ TEST_F(TestUnifySchemas, MoreSchemas) {
+  CheckUnify(fixed_size_list(int8(), 2), {list(int8()), large_list(int8())}, options);
+
+  options.promote_nested = true;
+  CheckUnify(list(int8()), {list(int16()), list(int32()), list(int64())}, options);
+  CheckUnify(fixed_size_list(int8(), 2),
+             {fixed_size_list(int16(), 2), list(int16()), list(int32()), list(int64())},
+             options);
+
+  auto ty = list(field("foo", int8(), /*nullable=*/false));
+  CheckUnifyAsymmetric(ty, list(int8()), list(field("foo", int8(), /*nullable=*/true)),
+                       options);
+  CheckUnifyAsymmetric(ty, list(field("bar", int16(), /*nullable=*/false)),
+                       list(field("foo", int16(), /*nullable=*/false)), options);
+}
+
+TEST_F(TestUnifySchemas, Map) {
+  auto options = Field::MergeOptions::Defaults();
+  options.promote_nested = true;
+  options.promote_numeric_width = true;
+
+  CheckUnify(map(int8(), int32()),
+             {map(int8(), int64()), map(int16(), int32()), map(int64(), int64())},
+             options);
+
+  // Do not test field names, since MapType intentionally ignores them in comparisons
+  // See ARROW-7173, ARROW-14999
+  auto ty = map(field("key", int8(), /*nullable=*/false),
+                field("value", int32(), /*nullable=*/false));
+  CheckUnify(ty, map(int8(), int32()),
+             map(field("key", int8(), /*nullable=*/true),
+                 field("value", int32(), /*nullable=*/true)),
+             options);
+  CheckUnify(ty,
+             map(field("key", int16(), /*nullable=*/false),
+                 field("value", int64(), /*nullable=*/false)),
+             map(field("key", int16(), /*nullable=*/false),
+                 field("value", int64(), /*nullable=*/false)),
+             options);
+}
+
+TEST_F(TestUnifySchemas, Struct) {
+  auto options = Field::MergeOptions::Defaults();
+  options.promote_nested = true;
+  options.promote_numeric_width = true;
+  options.promote_binary = true;
+
+  CheckUnify(struct_({}), struct_({field("a", int8())}), struct_({field("a", int8())}),
+             options);
+
+  CheckUnifyAsymmetric(struct_({field("b", utf8())}), struct_({field("a", int8())}),
+                       struct_({field("b", utf8()), field("a", int8())}), options);
+  CheckUnifyAsymmetric(struct_({field("a", int8())}), struct_({field("b", utf8())}),
+                       struct_({field("a", int8()), field("b", utf8())}), options);
+
+  CheckUnify(struct_({field("b", utf8())}), struct_({field("b", binary())}),
+             struct_({field("b", binary())}), options);
+
+  CheckUnifyAsymmetric(
+      struct_({field("a", int8()), field("b", utf8()), field("a", int64())}),
+      struct_({field("b", binary())}),
+      struct_({field("a", int8()), field("b", binary()), field("a", int64())}), options);
+
+  ASSERT_RAISES(
+      Invalid,
+      field("foo", struct_({field("a", int8()), field("b", utf8()), field("a", int64())}))
+          ->MergeWith(field("foo", struct_({field("a", int64())})), options));
```

Comment:
Since this is always tedious to write, I created https://github.com/apache/arrow/issues/36867

Reply:
For the record, that PR was merged, so you can use it now.